renderdoc/driver/d3d12/d3d12_pixelhistory.cpp

```diff
@@ -0,0 +1,401 @@
+/******************************************************************************
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2023 Baldur Karlsson
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+#include "driver/dxgi/dxgi_common.h"
+#include "d3d12_command_queue.h"
+#include "d3d12_debug.h"
+
+struct D3D12CopyPixelParams
+{
+  // The image being copied from
+  ID3D12Resource *srcImage;
+
+  // The source image format and format to use when copying. In most cases these are
+  // the same, but for some planar formats, the copy format will correspond to a single
+  // plane's typeless format, not the multi-plane format.
+  DXGI_FORMAT srcImageFormat;
+  DXGI_FORMAT copyFormat;
+  D3D12_RESOURCE_STATES srcImageState;
+
+  // Data about the pixel we want to copy
+  uint32_t x;
+  uint32_t y;
+  uint32_t mip;
+  uint32_t sample;
+  uint32_t planeSlice;
```

Comment:
I think you also need to account for array slice in here, it's currently being passed through as just 0 in the copy function.

Reply:
Good point. It would seem that the test I currently have (which is modeled after the Vulkan pixel history test) doesn't exercise that path. Hooking up what seems right, but we'll need to add more tests to verify it later.